lib/collection/src/shards/replica_set/clock_set.rs

```diff
@@ -0,0 +1,94 @@
+use std::cmp;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
+
+#[derive(Clone, Debug, Default)]
+pub struct ClockSet {
+    clocks: Vec<Arc<Clock>>,
+}
+
+impl ClockSet {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get the first available clock from the set, or create a new one.
+    pub fn get_clock(&mut self) -> ClockGuard {
+        for (id, clock) in self.clocks.iter().enumerate() {
+            if clock.lock() {
+                return ClockGuard::new(id, clock.clone());
+            }
+        }
+
+        let id = self.clocks.len();
+        let clock = Arc::new(Clock::new_locked());
+
+        self.clocks.push(clock.clone());
+
+        ClockGuard::new(id, clock)
+    }
+}
+
+#[derive(Debug)]
+pub struct ClockGuard {
+    id: usize,
+    clock: Arc<Clock>,
+}
+
+impl ClockGuard {
+    fn new(id: usize, clock: Arc<Clock>) -> Self {
+        Self { id, clock }
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    #[must_use = "new clock value must be used"]
+    pub fn tick_once(&mut self) -> u64 {
+        self.clock.tick_once()
+    }
+
+    pub fn advance_to(&mut self, new_tick: u64) -> u64 {
+        self.clock.advance_to(new_tick)
+    }
+}
+
+impl Drop for ClockGuard {
+    fn drop(&mut self) {
+        self.clock.release();
+    }
+}
+
+#[derive(Debug)]
+struct Clock {
+    clock: AtomicU64,
+    available: AtomicBool,
+}
+
+impl Clock {
+    fn new_locked() -> Self {
+        Self {
+            clock: AtomicU64::new(0),
+            available: AtomicBool::new(false),
+        }
+    }
+
+    #[must_use = "new clock value must be used"]
+    fn tick_once(&self) -> u64 {
+        self.clock.fetch_add(1, Ordering::Relaxed)
+    }
+
+    fn advance_to(&self, new_tick: u64) -> u64 {
```

Comment:
If I understood correctly, the main purpose of this function is to make sure, that no other thread will get ticks equal or less than `new_tick`. Return value might just create a confusion here, so I would propose to not return anything from the function

Reply:
If we want to return something here because we need in in the future, then I strongly recommend to rename this function to clarify what it actually returns.

Maybe something like `advance_to_get_next_tick`

Reply:
It returns current value of the clock (i.e., either `new_tick`, or whatever is in the clock, if it's already "newer" than `new_tick`). I don't use it anywhere ATM, but I was using it before at some point.

Reply:
> `advance_and_get_next_tick`

I tried to do something like this, but this also doesn't make much sense. Like, it returns "next tick"... but then the next call to `tick_once` returns the same tick? I agree, that it's just easier to not return anything.

Reply:
> I agree, that it's just easier to not return anything.

We can add it back later if we suddenly need it :+1: